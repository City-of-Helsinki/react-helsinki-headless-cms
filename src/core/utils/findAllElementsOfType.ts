import { getChildrenByType } from './getChildrenByType';
import { recursiveMap } from './recursiveMap';

/**
 * Find all elements of defined types recursively from a ReactNode.
 * @returns a list of elements of specified types that could be found from inside the ReactNode.
 */
export const findAllElementsOfType = (
  children: React.ReactNode,
  types: string[],
) => {
  const elementsOfType = [];
  recursiveMap(children, (child) => {
    elementsOfType.push(getChildrenByType(child, types));
    return child;
  });
  return elementsOfType.flat();
};
