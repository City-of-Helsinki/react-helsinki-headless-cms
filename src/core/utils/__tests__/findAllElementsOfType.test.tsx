import React from 'react';

import { findAllElementsOfType } from '../findAllElementsOfType';

describe('findAllElementsOfType', () => {
  const flattenedListItems = (
    <ul>
      <li>Should be included 1/2</li>
      <li>Should be included 2/2</li>
    </ul>
  );

  const nestedListItems = (
    <div>
      <ul>
        <li>Should be included 1/3</li>
      </ul>
      <ol>
        <li>Should be included 2/3</li>
        <li>Should be included 3/3</li>
      </ol>
    </div>
  );

  it.each([
    [flattenedListItems, ['li'], 2], // 2 x <li>
    [nestedListItems, ['li'], 3], // 3 x <li>
    [nestedListItems, ['ul', 'ol', 'li'], 5], // <ul>, <ol>, 3 x <li>
    [
      <div>
        <div>{nestedListItems}</div>
      </div>,
      ['li'],
      3, // 3 x <li>
    ],
    [
      <div>
        <div>{nestedListItems}</div>
        <div>
          <span>
            <img alt="should be included 1/2" src="#" />
          </span>
        </div>
        <img alt="should be included 2/2" src="#" />
        <div>
          <p>Not included</p>
        </div>
      </div>,
      ['img'],
      2, // 2 x <img>
    ],
  ])(
    'finds the elements of searched types #',
    (component, types, foundCount) => {
      const { children } = component.props;
      const result = findAllElementsOfType(children, types);
      expect(result).toHaveLength(foundCount);
      expect(result).toMatchSnapshot();
    },
  );
});
