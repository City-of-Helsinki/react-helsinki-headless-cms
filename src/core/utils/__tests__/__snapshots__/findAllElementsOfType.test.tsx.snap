// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findAllElementsOfType finds the elements of searched types # 1`] = `
[
  <li>
    Should be included 1/2
  </li>,
  <li>
    Should be included 2/2
  </li>,
]
`;

exports[`findAllElementsOfType finds the elements of searched types # 2`] = `
[
  <li>
    Should be included 1/3
  </li>,
  <li>
    Should be included 2/3
  </li>,
  <li>
    Should be included 3/3
  </li>,
]
`;

exports[`findAllElementsOfType finds the elements of searched types # 3`] = `
[
  <li>
    Should be included 1/3
  </li>,
  <ul>
    <li>
      Should be included 1/3
    </li>
  </ul>,
  <li>
    Should be included 2/3
  </li>,
  <li>
    Should be included 3/3
  </li>,
  <ol>
    <li>
      Should be included 2/3
    </li>
    <li>
      Should be included 3/3
    </li>
  </ol>,
]
`;

exports[`findAllElementsOfType finds the elements of searched types # 4`] = `
[
  <li>
    Should be included 1/3
  </li>,
  <li>
    Should be included 2/3
  </li>,
  <li>
    Should be included 3/3
  </li>,
]
`;

exports[`findAllElementsOfType finds the elements of searched types # 5`] = `
[
  <img
    alt="should be included 1/2"
    src="#"
  />,
  <img
    alt="should be included 2/2"
    src="#"
  />,
]
`;
